// class-converter/lib/decorators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultVal = exports.optional = exports.typed = exports.property = exports.serializeTarget = exports.afterSerialize = exports.beforeDeserialize = exports.deserialize = exports.serialize = void 0;
var store_1 = require("./store");
function serialize(serializer, disallowIgnore) {
    if (disallowIgnore === void 0) { disallowIgnore = false; }
    return function (target, propertyKey) {
        store_1.setStore(target, {
            key: propertyKey,
            serializer: serializer,
            disallowIgnoreSerializer: disallowIgnore,
        });
    };
}
exports.serialize = serialize;
function deserialize(deserializer, disallowIgnore) {
    if (disallowIgnore === void 0) { disallowIgnore = false; }
    return function (target, propertyKey) {
        store_1.setStore(target, {
            key: propertyKey,
            deserializer: deserializer,
            disallowIgnoreDeserializer: disallowIgnore,
        });
    };
}
exports.deserialize = deserialize;
function beforeDeserialize(beforeDeserializer, disallowIgnore) {
    if (disallowIgnore === void 0) { disallowIgnore = false; }
    return function (target, propertyKey) {
        store_1.setStore(target, {
            key: propertyKey,
            beforeDeserializer: beforeDeserializer,
            disallowIgnoreBeforeDeserializer: disallowIgnore,
        });
    };
}
exports.beforeDeserialize = beforeDeserialize;
function afterSerialize(afterSerializer, disallowIgnore) {
    if (disallowIgnore === void 0) { disallowIgnore = false; }
    return function (target, propertyKey) {
        store_1.setStore(target, {
            key: propertyKey,
            afterSerializer: afterSerializer,
            disallowIgnoreAfterSerializer: disallowIgnore,
        });
    };
}
exports.afterSerialize = afterSerialize;
function serializeTarget() {
    return function (target, propertyKey) {
        store_1.setStore(target, {
            key: propertyKey,
            serializeTarget: true,
        });
    };
}
exports.serializeTarget = serializeTarget;
function property(originalKey, targetClass, isOptional) {
    if (isOptional === void 0) { isOptional = false; }
    return function (target, propertyKey) {
        var config = {
            originalKey: originalKey || propertyKey,
            key: propertyKey,
        };
        if (targetClass) {
            config.targetClass = targetClass.prototype.constructor;
        }
        if (isOptional) {
            config.optional = isOptional;
        }
        store_1.setStore(target, config);
    };
}
exports.property = property;
function typed(targetClass) {
    return function (target, propertyKey) {
        store_1.setStore(target, {
            key: propertyKey,
            targetClass: targetClass,
        });
    };
}
exports.typed = typed;
function optional() {
    return function (target, propertyKey) {
        store_1.setStore(target, {
            key: propertyKey,
            optional: true,
        });
    };
}
exports.optional = optional;
function defaultVal(value) {
    return function (target, propertyKey) {
        store_1.setStore(target, {
            key: propertyKey,
            default: value,
        });
    };
}
exports.defaultVal = defaultVal;
